File: typescript/src/actions/evm/transfer/transfer.ts
import { createPublicClient, http, erc20Abi, parseUnits, Address, Chain, Transport } from "viem";

import { mapNetworkToChain } from "./utils.js";
import { EvmAccount, EvmSmartAccount } from "../../../accounts/evm/types.js";
import { CdpOpenApiClientType } from "../../../openapi-client/index.js";

import type {
  TransferResult,
  TransferExecutionStrategy,
  SmartAccountTransferOptions,
  AccountTransferOptions,
} from "./types.js";

/**
 * Transfer an amount of a token from an account to another account.
 *
 * @param apiClient - The client to use to send the transaction.
 * @param from - The account to send the transaction from.
 * @param transferArgs - The options for the transfer.
 * @param transferStrategy - The strategy to use to execute the transfer.
 * @returns The result of the transfer.
 */
export async function transfer<T extends EvmAccount | EvmSmartAccount>(
  apiClient: CdpOpenApiClientType,
  from: T,
  transferArgs: T extends EvmSmartAccount ? SmartAccountTransferOptions : AccountTransferOptions,
  transferStrategy: TransferExecutionStrategy<T>,
): Promise<TransferResult> {
  const publicClient = createPublicClient<Transport, Chain>({
    chain: mapNetworkToChain(transferArgs.network),
    transport: http(),
  });

  const to =
    typeof transferArgs.to === "string" ? transferArgs.to : (transferArgs.to.address as Address);

  const value = await (async () => {
    // user supplied a bigint. otherwise, we need to convert the amount to a bigint
    if (typeof transferArgs.amount !== "string") {
      return transferArgs.amount;
    }

    const decimals = await (async () => {
      if (transferArgs.token === "eth") {
        return 18;
      } else if (transferArgs.token === "usdc") {
        return 6;
      } else {
        return publicClient.readContract({
          address: transferArgs.token,
          abi: erc20Abi,
          functionName: "decimals",
          args: [],
        });
      }
    })();

    return parseUnits(transferArgs.amount, decimals);
  })();

  const transfer = {
    apiClient,
    from,
    to,
    value,
    token: transferArgs.token,
    network: transferArgs.network,
    paymasterUrl: "paymasterUrl" in transferArgs ? transferArgs.paymasterUrl : undefined,
  };

  const hash = await transferStrategy.executeTransfer(transfer);

  const result = await transferStrategy.waitForResult({
    apiClient,
    publicClient,
    from,
    hash,
    waitOptions: transferArgs.waitOptions,
  });

  return result;
}

File: python/cdp/actions/evm/transfer/transfer.py
from decimal import Decimal
from typing import TypeVar

from web3 import Web3

from cdp.actions.evm.transfer.constants import ERC20_ABI
from cdp.actions.evm.transfer.types import (
    TransferExecutionStrategy,
    TransferOptions,
    TransferResult,
)
from cdp.actions.evm.transfer.utils import get_chain_config
from cdp.api_clients import ApiClients
from cdp.evm_server_account import EvmServerAccount
from cdp.evm_smart_account import EvmSmartAccount

# Type for account
T = TypeVar("T", bound=EvmServerAccount | EvmSmartAccount)


async def transfer(
    api_clients: ApiClients,
    from_account: T,
    transfer_args: TransferOptions,
    transfer_strategy: TransferExecutionStrategy,
) -> TransferResult:
    """Transfer an amount of a token from an account to another account.

    Args:
        api_clients: The API clients to use to send the transaction
        from_account: The account to send the transaction from
        transfer_args: The options for the transfer
        transfer_strategy: The strategy to use to execute the transfer

    Returns:
        The result of the transfer

    """
    # Create a Web3 client for the specified network
    chain_config = get_chain_config(transfer_args.network)
    w3 = Web3(Web3.HTTPProvider(chain_config["rpc_url"]))

    # Determine the recipient address
    to_address = (
        transfer_args.to.address if hasattr(transfer_args.to, "address") else transfer_args.to
    )

    # Calculate the value to transfer
    value = await _calculate_value(w3, transfer_args)

    # Execute the transfer using the provided strategy
    tx_hash = await transfer_strategy.execute_transfer(
        api_clients=api_clients,
        from_account=from_account,
        transfer_args=transfer_args,
        to=to_address,
        value=value,
    )

    # Wait for the result of the transfer
    result = await transfer_strategy.wait_for_result(
        api_clients=api_clients, w3=w3, from_account=from_account, hash=tx_hash
    )

    return result


async def _calculate_value(w3: Web3, transfer_args: TransferOptions) -> int:
    """Calculate the value to transfer based on the transfer arguments.

    Args:
        w3: The Web3 client
        transfer_args: The transfer arguments

    Returns:
        The value to transfer

    """
    # If amount is already an integer, return it directly
    if isinstance(transfer_args.amount, int):
        return transfer_args.amount

    # Otherwise, convert the string amount to an integer based on the token's decimals
    amount_decimal = Decimal(transfer_args.amount)

    # Get the token decimals
    if transfer_args.token == "eth":
        decimals = 18
    elif transfer_args.token == "usdc":
        decimals = 6
    else:
        # For other tokens, get the decimals from the contract
        token_contract = w3.eth.contract(
            address=Web3.to_checksum_address(transfer_args.token), abi=ERC20_ABI
        )
        decimals = token_contract.functions.decimals().call()

    # Calculate the value with the right number of decimals
    value = int(amount_decimal * (10**decimals))
    return value
